File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int size = length(head);
         if(size == n){
            head = head.next;
            return head;
        }
        if(size == 1){
            return null;
        }
        if(size==2){
            head.next = null;
            return head;
        }
       
        ListNode t1 = head;
        ListNode t2 = head.next;
        ListNode t3 = head.next.next;
        int toIterate = size-n-1;
        for(int i = 1;i<=toIterate;i++){
            t1 = t2;
            t2 = t3;
            t3 = t3.next;
        }
        t1.next = t3;
        return head;
    }

    public int length(ListNode head){
        ListNode temp = head;
        int count = 0;
        while(temp!=null){
            count++;
            temp = temp.next;
        }
        return count;
    }
}

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* removeNthFromEnd(struct ListNode* head, int n) {
    struct ListNode *temp = head;
    struct ListNode *t1 = head;
    struct ListNode *prev = head;
    while(n>0){
        temp = temp->next;
        --n;
    }
    if(temp == NULL){
        return prev->next;
    }
    while(temp->next != NULL){
        prev = prev->next;
        temp= temp->next;
    }
    prev->next = prev->next->next;
    return head;
}